🚀 3단계 - 경로 조회 기능

hellonayeon opened the pull request:
## 실습 환경
* 경로 조회 도메인 (SubwayMap, SectionEdge) 은 샘플 코드를 사용

<img width="700" alt="image" src="https://github.com/next-step/atdd-subway-path/assets/43202607/e3f4a86f-7cd8-4967-ab17-24adf887e60d">


## 구현 기능
* 요구사항 설명에서 제공되는 추가된 요구사항을 기반으로 `경로 조회 기능`을 구현
* 인수 테스트 작성
* SubwayMap 도메인 단위테스트 작성


## 고민되는 부분
* 비즈니스 로직에서 null 체크 또는 유효성 검사하는 조건문이 많아지면 어떻게 하는게 좋을까요?
  * 예외처리나 null 방지를 위해서 조건문으로 검사하는 부분이 많아지면 가독성이 떨어진다고 느껴지는데요,
     관련해서 고민해보신 경험이나 관련 소스를 보신 경험이 있으시다면 공유 부탁드립니다!
  * 가독성이 떨어진다고 느끼는 것도 한 메서드 안에서 유효성 검사 포함 많은 로직을 담당하고 있어서 그런거라 생각해서 책임에 맞게 클래스를 분리하고, 또는 유효성 검사 클래스나 메서드를 따로 만드는 방법도 있을 것 같아요.!
  * 다른 분들은 어떻게 하시는지 궁금해서 여쭤봅니다 😃
```java
class Sections {

   ...

   private void processIfInsertedBetween(Section newSection) {
        Section connectedSection = getConnectedSection(newSection);

        if (newSection.isInsertedBetween(connectedSection)) { // 유효성 검사
            if (newSection.hasLoggerDistance(connectedSection)) { // 유효성 검사
                throw new CannotCreateSectionException();
            }

            Section dividedSection = getDividedSection(connectedSection, newSection);

            sections.add(dividedSection);
            sections.remove(connectedSection);
        }
    }

   ...
}
```

* 테스트 코드 Fixture 생성 시 테스트 환경을 위해 많은 데이터들을 생성하면, 어떤 객체들이 구성되어 있는지 한 눈에 안들어오더라구요.. 
  * Steps 클래스를 따로 빼고, RestAssured 요청 전송 클래스를 분리해봤는데도 가독성 좋게 리팩토링 되었다고 하기 애매하게 느껴졌어요.
  * 리뷰 하시면서 더 나은 개선방향에 대해 공유주시면 리팩토링 해보고자 합니다.

hellonayeon committed:
Docs: 인수 조건 도출

hellonayeon committed:
Docs: 구간 삭제 실패 케이스 인수 조건 수정
- 실패 메시지 검증 수정

hellonayeon committed:
Chore: 패키지 정리
- Assertions 검증 클래스 패키지 분리

hellonayeon committed:
Refactor: 노선 인수테스트 Steps 클래스 분리

hellonayeon committed:
Test: 지하철 경로 조회 성공 케이스 인수 테스트 작성
- 경로 조회에 필요한 Steps 정의
- 새로운 지하철역명, 노선명 등록
- 경로 응답 DTO 생성

hellonayeon committed:
Docs: 인수 조건 수정 (Simplify)

hellonayeon committed:
Test: 경로 조회 실패 케이스 인수 테스트 작성
- 경로 조회 실패 Exception 추가
- 인수 조건 수정 (출발역, 도착역 찾지 못한 경우 지하철 예외 메시지 검증)

hellonayeon committed:
Feat: 지하철 지도 도메인 추가 (샘플코드 참고)

hellonayeon committed:
Test: 지하철지도 도메인 단위 테스트 작성

hellonayeon committed:
Test: 인수 테스트 Fixture 수정

hellonayeon committed:
Feature: 경로 조회 서비스 레이어 추가

hellonayeon committed:
Feature: 경로 조회 컨트롤러 추가

hellonayeon committed:
Refactor: 사용하지 않는 메서드 제거

hiblue02 commented on src/main/java/subway/domain/Sections.java:
```diff
@@ -22,11 +22,19 @@ public class Sections {
     }
 
+    public List<Section> get() {
+        return sections;
```
외부의 영향을 줄이기 위해, unmodifiable로 반환해보면 어떨까요?

hiblue02 commented:
> ## 고민되는 부분
> * 비즈니스 로직에서 null 체크 또는 유효성 검사하는 조건문이 많아지면 어떻게 하는게 좋을까요?
> * 가독성이 떨어진다고 느끼는 것도 한 메서드 안에서 유효성 검사 포함 많은 로직을 담당하고 있어서 그런거라 생각해서 책임에 맞게 클래스를 분리하고, 또는 유효성 검사 클래스나 메서드를 따로 만드는 방법도 있을 것 같아요.!

저는 검증을 담당하는 객체를 만드는 방법으로 풀어봤어요. 😄 
feat. 책임 채인 패턴, 전략 패턴
- 책임 채인 패턴: https://refactoring.guru/ko/design-patterns/chain-of-responsibility
- 전략 패턴 : https://refactoring.guru/ko/design-patterns/strategy

Null 검증은 BeanValidation을 활용하기도 합니다. 
~~~
    @NotNull(message = "출발역은 반드시 입력해야 합니다.")
    private Long source;

    @NotNull(message = "도착역은 반드시 입력해야 합니다.")
    private Long target;
~~~

hiblue02 commented on src/main/java/subway/service/PathService.java:
```diff
@@ -0,0 +1,29 @@
+    public Path findPath(Long sourceId, Long targetId) {
+        if (sourceId.equals(targetId)) {
+            throw new CannotFindPath();
+        }
```
Validation을 SubwayMap에서 하면 응집도가 높아질 것 같아요.

hiblue02 commented on src/test/java/subway/acceptance/PathAcceptanceTest.java:
```diff
@@ -0,0 +1,121 @@
+        // Then
+        경로_조회_예외_검증(조회한_경로.as(ErrorResponse.class),
+            SubwayErrorCode.STATION_NOT_FOUND.getMessage());
+    }
```
실패 케이스도 꼼꼼히 테스트 해주셨네요. :100:

hiblue02 commented on src/test/java/subway/acceptance/PathAcceptanceTest.java:
```diff
@@ -0,0 +1,121 @@
+import subway.exception.error.SubwayErrorCode;
+
+@DisplayName("경로 조회 기능")
+@DirtiesContext(classMode = ClassMode.AFTER_EACH_TEST_METHOD)
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+public class PathAcceptanceTest {
```
AcceptanceTest를 상속해보면 어떨까요? :smile:

hiblue02 commented:
> 테스트 코드 Fixture 생성 시 테스트 환경을 위해 많은 데이터들을 생성하면, 어떤 객체들이 구성되어 있는지 한 눈에 안들어오더라구요..
> * Steps 클래스를 따로 빼고, RestAssured 요청 전송 클래스를 분리해봤는데도 가독성 좋게 리팩토링 되었다고 하기 애매하게 느껴졌어요.

1. Steps와 RestAssuredClient, SubwayFactory의 역할이 좀더 명확하게 구분되어야 할 것 같아요. Steps가 텍스처를 생성하기도 하고, RestAssuredClient를 통해 요청을 전송하기도 하네요. 그리고 SubwayFactory의 역할이 거의 없네요. 😭 
2. 저는 텍스트 Fixture는 테스트를 하는 곳과 가장 가깝게 있어야 한다고 생각해요. 다른 객체에서 왔다갔다하면서 봐야해서 한눈에 들어오지 않더라구요. 
3. 텍스트 Fixture 생성할 때는 가독성을 위해 Builder가 아닌 기본 생성자를 써보거나, 지하철_생성 메소드를 추출해보면 어떨까요?
~~~ 
교대역 = 지하철역_생성_요청("교대역").jsonPath().getLong("id");
강남역 = 지하철역_생성_요청("강남역").jsonPath().getLong("id");
~~~

hiblue02 reviewed (changes requested):
3단계도 꼼꼼히 구현해주셨네요! 👍 
코멘트 남겼으니 확인해주시고, 다시 리뷰요청 주세요!

hellonayeon committed:
Refactor: 인수 테스트 리팩토링
- 책임이 명확하지 않은 클래스 제거 (RestAssuredClient, SubwayNameFactory, LineRequestFactory)
- 지하철 도메인 관련 요청을 보내는 클래스를 추가: SubwayClient
- 아이디 기반 요청/응답 검증을 위해 멤버변수 (Long 타입의 역이름, 노선이름) 추가
- 인수 테스트 설정 클래스 상속 (AcceptanceTest)

hellonayeon committed:
Refactor: Optional Empty 체크 stream() 체이닝 메서드로 수정 (orThrowElse())

hellonayeon committed:
Refactor: 구간 추가 메서드 분리
- 기존 구간 사이에 새로운 역을 추가하는 경우
- 기존 구간 양 끝에 새로운 역을 추가하는 경우

hellonayeon committed:
Refactor: 경로조회요청 파라미터 Validator Chain 추가

hellonayeon committed:
Refactor: 경로 조회 Request DTO Bean Validation 추가
- BindException 핸들러 추가
- 경로 조회 요청 @RequestParam -> @ModelAttribute 바인딩으로 수정

hellonayeon committed:
Refactor: 불필요한 Sections 생성자 제거

hellonayeon committed:
Refactor: Sections add 메서드 분리

hellonayeon commented:
## 리팩터링 내용
* 인수 테스트 수정 및 불변 컬렉션 리턴하도록 수정 ([f5d7824](https://github.com/next-step/atdd-subway-path/pull/688/commits/f5d7824e865ea6d0d0a2c662c2259ea099f8c3db))
  * Fixture, Steps 수정 (불필요한 RestClient, Factory 제거)
  * Steps 에는 지하철 api 로 요청을 보내는 메서드만 가지도록 수정
  * 테스트 시 필요한 데이터는 각 인수테스트의 `@BeforeEach`에서 초기화하도록 수정

* 검증 로직 리팩토링
  * 경로 조회 요청 파라미터 검증을 위한 책임 연쇄 패턴 추가
  * 요청 파라미터 Bean Validation, BindExceptionHandler 추가

* Sections add 메서드 분리
  * 구간 추가하는 기능에서 (삭제도 마찬가지지만) 기존 구간의 사이에 새로운 역이 추가되는 경우, 기존 구간의 양끝에 새로운 역이 추가되는 경우 와 같이 추가 방식을 두 가지로 분리해서 생각했는데요
  * 전략패턴을 통해 가독성을 높일 수 있지 않을까 고민해봤는데 아래와 같은 고민으로 메서드 분리만 했어요..
    * _일급 컬렉션 객체에서 전략 패턴 객체와 연관관계를 맺어도 되는건지? (일급 컬렉션은 컬렉션 객체만을 단일 필드로 갖고있는 클래스인데)_

## 느낀점
* 책임 연쇄 패턴이 spring filter 나 aop 와 유사하다고 느껴졌는데, filter / aop / chain of resposibility pattern 이들을 어떨때 사용하는게 좋을까 고민해볼 수 있는 시간이었어요.
  * filter: 컨트롤러 수행 전 후로 http request, response 의 가공이나 인증처리가 필요할때 사용
  * aop: logging, 자원 초기화 또는 반환과 같이 비즈니스 로직에서 일관된 처리가 필요할때 사용
  * chain of responsibility pattern: 데이터(dto, domain) 기반으로 검증이나 선행 처리가 필요할때 사용

hellonayeon commented on src/main/java/subway/controller/PathController.java:
```diff
@@ -0,0 +1,32 @@
+
+    private final PathService pathService;
+
+    @GetMapping
+    public ResponseEntity<PathResponse> findPath(@Valid PathRequest request) {
+        PathValidator validator = new PathSourceTargetEqualsValidator(null); // validator chaining 가능
+        validator.validate(request);
+
+        Path path = pathService.findPath(request);
+        return ResponseEntity.ok(PathResponse.from(path));
+    }
```
service 계층에서 controller 계층으로 검증 로직을 단순히 이동한게 아닌가.. 
만약 다른 validator 가 추가된다면 책임 체이닝을 위해 controller 의 변경이 불가피할텐데 - 라는 생각이 들었습니당
하지만 새로운 검증 로직이 생긴다면 Validator 를 추가해가며 분리해서 관리할 수 있다는 장점이 있을 것 같아요.

아직 제가 책임연쇄패턴에 대해 추상적인 개념만을 갖고 있어서 그런게 아닐까 싶어요, 
혹시 리뷰하시면서 잘못된 부분이나 더 나은 방식이 있다면 공유 부탁드려요! 남은 미션들에서 적용해보겠습니다아!

hiblue02 replied:
지금은 검증 로직이 1개 밖에 없어서 그렇게 느껴지지만, 이전보다 훨씬 유연한 설계로 보여요! 
controller에서 validator를 직접 호출하고 있는데 애노테이션을 이용해 적용하는 방법도 있어요. [링크](https://tecoble.techcourse.co.kr/post/2021-06-21-custom-annotation/)를 참고해보세요.

hellonayeon commented on src/main/java/subway/domain/Sections.java:
```diff
@@ -22,31 +22,37 @@ public class Sections {
     }
 
     public void add(Section section) {
-        processIfInsertedBetween(section);
+        removeSectionForAdd(section);
         sections.add(section);
     }
 
```
* SectionAddStrategy
  * SectionBetweenAddStrategy
  * SectionEndAddStrategy

구간을 추가하는 전략을 나눈다면, 전략에 대한 연관관계는 일급컬렉션 객체에서 갖고 있어도 되는지

hiblue02 replied:
개인적으론! 전략을 검증하기 위한 데이터(?)를 일급컬렉션에서 갖고 있기 때문에, 일급컬랙션에 있어도 된다고 생각해요.  :smile:

hellonayeon commented on src/main/java/subway/domain/Sections.java:
```diff
@@ -22,31 +22,37 @@ public class Sections {
-    private void processIfInsertedBetween(Section newSection) {
-        Section connectedSection = getConnectedSection(newSection);
-
-        if (newSection.isInsertedBetween(connectedSection)) {
-            if (newSection.hasLoggerDistance(connectedSection)) {
-                throw new CannotCreateSectionException();
-            }
```
또는 어떻게 책임연쇄패턴을 적용하면 좋을지 방법이 구체적으로 떠오르지 않았습니다 😭

hiblue02 replied:
이렇게 표현해보면, 어떨까요? 
그리고 지금은 로직에 검증하는 부분이 여기저기 흩어져 있어요! 그래서 책임 연쇄 패턴 적용이 어려워보여요. 검증을 하나로 모을건지, 메소드에 나눠서 적용할지는  설계하기에 따라 다른거니 이번 단계에서는 책임 연쇄 패턴으로 풀 수도 있다 정도로만 인지하고 넘어가보면 어떨까요? :smile:
~~~
if (newSection.isNotInBetween(connectedSection)) {
    throw new CannotCreateSectionException();
}
if (newSection.hasLoggerDistance(connectedSection)) {
     throw new CannotCreateSectionException();          
}
~~~

hiblue02 commented on src/test/java/subway/acceptance/LineAcceptanceTest.java:
```diff
@@ -1,193 +1,136 @@
-    }
-
-    private static final String linePath = "/lines";
+class LineAcceptanceTest extends AcceptanceTest {
```
리뷰 반영 :+1:

hiblue02 commented on src/test/java/subway/acceptance/LineAcceptanceTest.java:
```diff
@@ -1,193 +1,136 @@
+        Long 광교역 = 지하철역_생성_요청(new StationRequest("광교역")).jsonPath().getLong("id");
 
-        assertThat(response.statusCode()).isEqualTo(HttpStatus.OK.value());
+        // When
+        Long 신분당선 = 노선_생성_요청(new LineRequest("신분당선", "bg-red-600", 강남역, 광교역, 30L)).jsonPath().getLong("id");
 
-        List<String> lineNames = response.jsonPath().getList("name", String.class);
-        assertThat(lineNames.size()).isEqualTo(1);
-        assertThat(lineNames).contains(신분당선);
+        // Then
+        ExtractableResponse<Response> 노선_조회_응답 = 노선_조회_요청(신분당선);
+        노선_응답_성공_검증(노선_조회_응답, HttpStatus.OK, 30L, List.of(강남역, 광교역));
```
인수테스트 코드가 보기 좋아졌네요! :clap::clap:

hiblue02 commented on src/main/java/subway/dto/PathRequest.java:
```diff
@@ -0,0 +1,18 @@
+@Getter @Setter
+public class PathRequest {
+
+    @NotNull(message = "출발역은 반드시 입력해야합니다.")
+    private Long source;
+
+    @NotNull(message = "도착역은 반드시 입력해야합니다.")
+    private Long target;
```
BeanValidation 사용 :clap::clap:

hiblue02 commented on src/main/java/subway/validation/PathValidator.java:
```diff
@@ -0,0 +1,19 @@
+
+public abstract class PathValidator {
+
+    private PathValidator nextValidator;
+
+    public PathValidator(PathValidator nextValidator) {
+        this.nextValidator = nextValidator;
+    }
+
+    public void validate(PathRequest request) {
+        if (nextValidator != null) {
+            nextValidator.validate(request);
+        }
+    }
```
책임 체인 패턴 사용 :100:

hiblue02 reviewed (approved):
나연님! 미션 수행하느라 고생 많으셨습니다. 💯 
코멘트 남겼으니 확인해주시고, 다음 미션도 화이팅입니다! 🔥
